frontend/apps/reader/modules/readerannotation.lua

```diff
@@ -0,0 +1,158 @@
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+
+local ReaderAnnotation = WidgetContainer:extend{}
+
+function ReaderAnnotation:buildAnnotation(bm, highlights, update_pageno)
+    -- bm - corresponding bookmark, highlights - all highlights
+    local hl, pageno = self:getHighlightByDatetime(highlights, bm.datetime)
+    if update_pageno then -- cannot be done in onReadSettings()
+        pageno = self.ui.paging and bm.page or self.document:getPageFromXPointer(bm.page)
+    end
+    if not hl then -- page bookmark or orphaned bookmark
+        hl = {}
+        if bm.highlighted then -- orphaned bookmark
+            hl.drawer = self.view.highlight.saved_drawer
+            if self.ui.paging then
+                if bm.pos0.page == bm.pos1.page then
+                    hl.pboxes = self.document:getPageBoxesFromPositions(bm.page, bm.pos0, bm.pos1)
+                else -- multi-page highlight, restore the first box only
+                    hl.pboxes = self.document:getPageBoxesFromPositions(bm.page, bm.pos0, bm.pos0)
```

Comment:
Can this (calling getPageBoxesFromPositions) be done in onReadSettings() ? (cf your comment on line 8).

Reply:
Actually I had no problems with doing it all (incl. line 8) in onReadSettings, but I saw your
https://github.com/koreader/koreader/blob/424fa9b6fd37ac09de80ff9ebbc96c33ed0da1d7/frontend/apps/reader/modules/readerbookmark.lua#L368-L369

Reply:
That's not _my_ comment, it's from 10 years ago, before I was born to KOReader :) #1270
But I agree with that comment, at least with CREDocuments: it's safer/necessary to delay all xpointers and co. stuff to after the document is renderered.